### DISCUSSION TITLE
Waves are not turning yellow as track plays

### DISCUSSION THREAD

**Discussion by imperium11** [naomiaro/waveform-playlist#194] [Q&A] [answered]
Hi everyone,

I am trying to use Waveform with React and i am very new to this library. When my component renders i am able to create a WaveformPlaylist and afterwards load audio files. But when i play the song i don't see a horizontal line following the track and filling in the wave forms.

Here is my code:

`useEffect(() => {
    setPlayList(WaveformPlaylist({
      samplesPerPixel: 1000,
      mono: true,
      waveHeight: 150,
      barWidth: 3,
      barGap: 1,
      timescale: true,
      container: document.getElementById('editor'),
      state: 'cursor',
      colors: {
        waveOutlineColor: 'grey',
        timeColor: 'purple',
        fadeColor: 'grey',
      },
      controls: {
        show: true,
        width: 130,
        widgets: {
          stereoPan: false,
          collapse: false
        }
      },
      zoomLevels: [1000],
    }));
  }, []);
`

`
playlist.load([{
          src: result.data.url,
          name: 'Track ${count}'
        }]);
`

And then i just use `playlist.getEventEmitter().emit` for my buttons:
`
<Pause onClick={() => { playlist.getEventEmitter().emit('pause'); }}></Pause>
            <Play onClick={() => { playlist.getEventEmitter().emit('play'); }}></Play>
            <Stop onClick={() => { playlist.getEventEmitter().emit('stop'); }}></Stop>
            <Rewind onClick={() => { playlist.getEventEmitter().emit('rewind'); }}></Rewind>
            <FastForward onClick={() => { playlist.getEventEmitter().emit('fastforward'); }}></FastForward>
`

And here is a screen shot of what i have so far:
![Screenshot from 2022-02-16 16-41-25](https://user-images.githubusercontent.com/81498700/154394421-eb4bf260-b438-4f3e-9661-94f3e925bda3.png)

I was expecting something like this:
![Screenshot from 2022-02-16 17-05-26](https://user-images.githubusercontent.com/81498700/154397035-57494afb-af80-429b-9349-bc44bda9d618.png)
Any help or advice would be greatly appreciated!

**Comment by naomiaro** [comment, marked as answer]
Hello, did you include the stylesheet?

https://unpkg.com/browse/waveform-playlist@4.2.1/styles/

it should be variable --wp-channel-progress-color

**Reply by imperium11** [reply]
oh i had no idea about the css file!!!!! thank you!

**Reply by naomiaro** [reply]
Glad it's working for you now!

**Reply by imperium11** [reply]
Hi, i also have another question, i can't get the x icon (remove track) to display. Why do you think that is? Also when we have tracks that have different lengths, why does the audio stop after the track that has the smallest length ends? for instance:

![Screenshot from 2022-02-17 00-15-45](https://user-images.githubusercontent.com/81498700/154455278-70e7012c-6750-4a15-ab4c-321e478ee1f9.png)
 
Thank you!

**Reply by naomiaro** [reply]
The css there is using some icons from font awesome. I included a custom kit on the demos:

https://github.com/naomiaro/waveform-playlist/blob/main/examples/basic-html/index.html#L7

**Reply by naomiaro** [reply]
Hmm I don't know about this length issue, I wonder if there is a bug there, I'll check it out!

**Reply by naomiaro** [reply]
Oh yeah I see this issue on the example here: https://naomiaro.github.io/waveform-playlist/web-audio-editor.html

It wasn't like that before, I'll see if something has come up to cause a problem.

**Reply by naomiaro** [reply]
Oh yeah I think I know what the issue is, thanks for reporting this.

**Reply by naomiaro** [reply]
I've published v4.2.2 which should resolve this issue.

**Reply by imperium11** [reply]
Thank you v4.2.2 works great! I put the font awesome link as a script but for some reason the `X` icon is still not showing. i am using all the css styling in my main.css file. Is there a way for me to put my own icon there somehow?

**Reply by naomiaro** [reply]
Is it in the `<head>` of the page? I think it needs to be there to work. Otherwise I'm not sure why it wouldn't show up.

**Reply by naomiaro** [reply]
If you want custom icons, you would have to override bits of the CSS to include them.

**Reply by naomiaro** [reply]
https://fontawesome.com/ It's best to create your own kit here of the icons you use if you're expecting a lot of views. I'm on the free plan which is open atm and runs the risk of running out of page views.

**Reply by naomiaro** [reply]
I should prob make a note of this somewhere and create some override options..

**Reply by imperium11** [reply]
ok thanks i will try that and thank you for being so helpful!

**Reply by naomiaro** [reply]
Thanks 😄 

I did try another way to add the icons: https://github.com/naomiaro/waveform-playlist/blob/main/examples/basic-nextjs/pages/_app.js#L3 I used the fontawesome free npm package and just included their all.css for now.

**Reply by imperium11** [reply]
Hi again, sorry to keep the discussion going here. I am able to use `ee.emit()` onClick functions without using the toggleActive function. What is the purpouse of the toggleActive function exactly. I tried to invoke it with `this` but the node is always `undefined`

**Reply by naomiaro** [reply]
oh you won't need to use that one with React. I was just using that to track the state of various Bootstrap buttons I was using in this simple jQuery example.

**Reply by imperium11** [reply]
What you guys have created is simply amazing. It has literally saved me so much time i can't thank you and all the people that have worked on this project enough! I pretty much have all the functionality of Full Waveform Editor example you guys have. Are there ways to add more effects to tracks like Reverb etc?

**Reply by naomiaro** [reply]
Thank-you! 😄 . Yes you can do that now! You can add Tone.js effects to the playlist now!

Tone.js reverb:
https://tonejs.github.io/docs/r13/Reverb

Demo in action
https://naomiaro.github.io/waveform-playlist/effects.html

Code for the demo:
https://github.com/naomiaro/waveform-playlist/blob/main/ghpages/js/effects.js#L53

You'll have to just modify the example code slightly to swap out the Tone.js effect. You can always chain multiple effects.

**Reply by imperium11** [reply]
Thank you for all of the links i will look into implementing the Reverb functionality this weekend. For now my biggest priority is being able to flatten all the tracks into 1 audio file and download it. I tried to follow the examples on `emitter.js` file. Would you be able to tell me what the process is when it comes to being able to download the edited audio tracks. I have a download button and want to be able to download the the audio file without displaying a link `<a></a>` Do i first need to convert the playlist into something else? i tried somethings like `URL.createObjectURL(playlist)` and using `playlist.getEventEmitter().emit('startaudiorendering', 'wav')`but i am not too sure where i am going wrong. Im only getting errors. Thanks again for all the help and info

**Reply by imperium11** [reply]
i guess i need to convert all the audio chunks into a Blob which would flatten it?

**Reply by naomiaro** [reply]
https://github.com/naomiaro/waveform-playlist/blob/main/ghpages/js/web-audio-editor.js#L81 You need to make sure to initialise the recorder. All functionality should be there already.

**Reply by naomiaro** [reply]
https://github.com/naomiaro/waveform-playlist/blob/main/ghpages/js/emitter.js#L225

**Reply by imperium11** [reply]
`playlist.load([{
          src: result.data.url,
          name: Track #1
        }]).then(() => {
          playlist.initExporter();
        });`

I also added the `playlist.getEventEmitter().emit('startaudiorendering', 'wav');` to my download (button) Onclick event handler but nothing seems to be happening (I am not getting any errors). Do i still need to convert all the files into a Blob and go into that process?

**Reply by imperium11** [reply]
This returns `undefined` within my `.then` function -> `console.log('initExporter', playlist.initExporter());`

**Reply by naomiaro** [reply]
https://github.com/naomiaro/waveform-playlist/blob/main/examples/basic-nextjs/pages/index.js

I added a quick download example here. I've added the packages `events`/`save-as` to do this.

**Reply by imperium11** [reply]
I cannot stress enough how grateful i am for your help. Thank you so much.

**Reply by imperium11** [reply]
Hi, is there a way to add a `reverb` effect with the click of a button on a specific track even though the track has been loaded with `playlist.load`. Are there ways to alter the playlist so i can add effects after the track is already on the editor (just like Fade-in fade-out but for Reverb)
